hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/gas/SystemContractGasCalculator.java

```diff
@@ -57,44 +59,60 @@ public long gasRequirement(
-        final var priceInTinybars = Math.max(minimumPriceInTinybars, nominalPriceInTinybars);
-        return asGasRequirement(priceInTinybars);
+    public long gasRequirementWithTinyCents(
+            @NonNull final TransactionBody body, @NonNull final AccountID payer, final long minimumPriceInTinyCents) {
+        // If not enabled, make the calculation using the old method.
+        if (!tinybarValues.isGasPrecisionLossFixEnabled()) {
+            return gasRequirementOldWithPrecisionLoss(body, payer, minimumPriceInTinyCents);
+        }
+
+        final var nominalPriceInTinyBars = feeCalculator.applyAsLong(body, payer);
+        // For the rare cases where nominalPriceInTinyBars > minimumPriceInTinyCents:
+        // Precision loss may occur as we convert between tinyBars and tinyCents, but it is typically negligible.
+        // The minimal nominal price is > 1e6 tinyCents, ensuring enough precision.
+        // In most cases, the gas difference is zero.
+        // In scenarios where we compare significant price fluctuations (200x, 100x), the gas difference should still be
+        // unlikely to exceed 0 units.
+        final var priceInTinyCents =
+                Math.max(minimumPriceInTinyCents, tinybarValues.asTinyCents(nominalPriceInTinyBars));
+
+        return gasRequirementFromTinyCents(priceInTinyCents, tinybarValues.childTransactionTinyCentsGasPrice());
     }
 
     /**
-     * Given a transaction body whose, returns the gas requirement for the transaction to be
-     * dispatched using the gas price of the top-level HAPI operation and no markup.
+     * Returns the gas price for the top-level HAPI operation.
      *
-     * @param body the transaction body to be dispatched
-     * @param payer the payer of the transaction
-     * @return the gas requirement for the transaction to be dispatched
+     * @return the gas price for the top-level HAPI operation in tinyBars.
      */
-    public long topLevelGasRequirement(@NonNull final TransactionBody body, @NonNull final AccountID payer) {
-        return feeCalculator.applyAsLong(body, payer) / tinybarValues.topLevelTinybarGasPrice();
+    public long topLevelGasPriceInTinyBars() {
+        return tinybarValues.topLevelTinybarGasPriceFullPrecision();
     }
 
     /**
-     * Returns the gas price for the top-level HAPI operation.
+     * Estimates the gas requirement for a view operation.
+     * The minimum gas requirement is 100 units.
+     * For all view operations, the gas requirement is determined using the canonical gas value
+     * for the TOKEN_INFO dispatch type, as specified in the canonical-prices.json.
+     * The TOKEN_INFO operation is representative of view operations.
      *
-     * @return the gas price for the top-level HAPI operation
+     * @return the gas requirement for a view operation
      */
-    public long topLevelGasPrice() {
-        return tinybarValues.topLevelTinybarGasPrice();
+    public long viewGasRequirement() {
```

Comment:
(Now that we have `viewGasRequirement` it seems like `canonicalGasRequirement` should have been called something different that is directly related to and opposite from "view".  But it has a bunch of callers already; not worth changing; just something to remember. Maybe the Javadoc down there at `canonicalGasRequirement` could mention explicitly that it's for the operations that (can) change state?)

Reply:
I've added "Used for non-query operations."